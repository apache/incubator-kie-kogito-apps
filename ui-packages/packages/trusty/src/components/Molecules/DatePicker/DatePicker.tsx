import React, { useEffect, useRef } from 'react';
import { InputGroup, TextInput } from '@patternfly/react-core';
import flatpickr from 'flatpickr';
import { Instance as flatpickrInstance } from 'flatpickr/dist/types/instance';
import leftArrow from '../../../../static/images/leftArrow.svg';
import rightArrow from '../../../../static/images/rightArrow.svg';
import './DatePicker.scss';

type DatePickerProps = {
  fromDate?: string;
  id: string;
  label?: string;
  minDate?: string;
  maxDate?: string;
  onDateUpdate: (date: string) => void;
  value?: string;
};

const DatePicker = (props: DatePickerProps) => {
  const { fromDate, minDate, maxDate, value, onDateUpdate, id, label } = props;
  const datePicker = useRef<HTMLInputElement>();

  useEffect(() => {
    let calendar: flatpickrInstance;
    const onChange = (selectedDates: Date[], dateStr: string) => {
      onDateUpdate(dateStr);
    };
    if (datePicker.current) {
      calendar = flatpickr(datePicker.current, {
        allowInput: true,
        altInput: true,
        altFormat: 'F j, Y',
        dateFormat: 'Y-m-d',
        defaultDate: value,
        minDate,
        maxDate,
        monthSelectorType: 'static',
        onChange,
        prevArrow: `<img alt="Prev" src="${leftArrow}">`,
        nextArrow: `<img alt="Next" src="${rightArrow}">`,
        static: true
      });
    }
    return () => {
      calendar.destroy();
    };
  }, [fromDate, maxDate, onDateUpdate, minDate, value, id]);

  return (
    <InputGroup>
      <TextInput
        name={id}
        id={id}
        type="date"
        aria-label={label}
        ref={datePicker}
      />
    </InputGroup>
  );
};

export default DatePicker;

export const datePickerSetup = () => {
  flatpickr.l10ns.en.weekdays.shorthand.forEach((day, index, daysArray) => {
    if (daysArray[index] === 'Thu' || daysArray[index] === 'Th') {
      daysArray[index] = 'Th';
    } else if (daysArray[index] === 'Sat' || daysArray[index] === 'Sa') {
      daysArray[index] = 'Sa';
    } else {
      daysArray[index] = daysArray[index].charAt(0);
    }
  });
};
