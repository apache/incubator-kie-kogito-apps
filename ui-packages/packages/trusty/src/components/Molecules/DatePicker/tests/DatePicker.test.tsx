import React from 'react';
import DatePicker from '../DatePicker';
import { mount } from 'enzyme';
import { render, screen, waitFor } from '@testing-library/react';

describe('DatePicker', () => {
  test('renders correctly', () => {
    const props = {
      id: 'datepicker',
      onDateUpdate: jest.fn()
    };
    const wrapper = mount(<DatePicker {...props} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('input#datepicker')).toHaveLength(1);
  });

  test('displays dates in the correct format', async () => {
    const props = {
      id: 'datepicker',
      value: '2020-06-01',
      onDateUpdate: jest.fn()
    };

    render(<DatePicker {...props} />);
    await waitFor(() =>
      expect(screen.getByDisplayValue('June 1, 2020')).toBeTruthy()
    );
  });
});
