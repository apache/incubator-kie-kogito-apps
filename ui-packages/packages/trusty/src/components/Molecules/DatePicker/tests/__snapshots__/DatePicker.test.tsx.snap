// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker renders correctly 1`] = `
<DatePicker
  id="datepicker"
  onDateUpdate={[MockFunction]}
>
  <Component>
    <div
      className="pf-c-input-group"
    >
      <ForwardRef
        id="datepicker"
        name="datepicker"
        type="date"
      >
        <TextInputBase
          aria-label={null}
          className=""
          id="datepicker"
          innerRef={
            Object {
              "current": <input
                aria-invalid="false"
                class="pf-c-form-control flatpickr-input"
                id="datepicker"
                name="datepicker"
                type="hidden"
                value=""
              />,
            }
          }
          isDisabled={false}
          isReadOnly={false}
          isRequired={false}
          name="datepicker"
          onChange={[Function]}
          type="date"
          validated="default"
        >
          <input
            aria-invalid={false}
            aria-label={null}
            className="pf-c-form-control"
            disabled={false}
            id="datepicker"
            name="datepicker"
            onChange={[Function]}
            readOnly={false}
            required={false}
            type="date"
          />
        </TextInputBase>
      </ForwardRef>
    </div>
  </Component>
</DatePicker>
`;
